File: usecases/name_usecase.py
from datetime import datetime
import uuid
from fastapi import HTTPException

from constants.error_response import ERROR_RESPONSES
from dtos.request.name_req_dto import NameReqDTO
from entities.user_entity import UserEntity
from repositories.name_repository import NameRepository


class NameUseCase:
    @staticmethod
    def get_names():
        # 이름을 레포지토리를 사용해서 가져오기
        try:
            name_list = NameRepository.get_names()
            return name_list  # 핸들러에서 NameListResDTO로 변환
        except Exception as e:
            raise HTTPException(
                status_code=500, detail=ERROR_RESPONSES["SERVER_ERROR"]
            ) from e

    @staticmethod
    def delete_name_by_index(index: int):
        name_list = NameRepository.get_names()
        if index < 0 or index >= len(name_list):
            raise HTTPException(
                status_code=400, detail=ERROR_RESPONSES["INVALID_INDEX"]
            )
        try:
            deleted_name = name_list[index]
            NameRepository.delete_name_by_index(index)
            return deleted_name
        except Exception as e:
            raise HTTPException(
                status_code=500, detail=ERROR_RESPONSES["SERVER_ERROR"]
            ) from e

    @staticmethod
    def delete_name_by_name(name: str):
        try:
            deleted_name = NameRepository.delete_name_by_name(name)
            # 해당 이름을 찾을 수 없는 경우
            if not deleted_name:
                raise HTTPException(
                    status_code=404, detail=ERROR_RESPONSES["USER_NOT_FOUND"]
                )
            return deleted_name
        except Exception as e:
            raise HTTPException(
                status_code=500, detail=ERROR_RESPONSES["SERVER_ERROR"]
            ) from e

    @staticmethod
    def get_name_by_name(name: str):
        name_list = NameRepository.get_names()
        if name not in [item.name for item in name_list]:
            raise HTTPException(
                status_code=404, detail=ERROR_RESPONSES["USER_NOT_FOUND"]
            )
        try:
            user = NameRepository.find_by_name(name)
            return user  # 핸들러에서 NameResDTO로 변환
        except Exception as e:
            raise HTTPException(
                status_code=500, detail=ERROR_RESPONSES["SERVER_ERROR"]
            ) from e

File: UseCases/NameUseCase.cs
using BerryNameApi.DTO.Response;
using BerryNameApi.Entities;
using BerryNameApi.Repositories;
using BerryNameApi.Utils;
using blueberry_homework_dotnet.UseCases;
using blueberry_homework_dotnet.Utils;

namespace BerryNameApi.UseCases
{
    public class NameUseCase
    {
        private readonly NameRepository _repository;

        public NameUseCase(NameRepository repository)
        {
            _repository = repository;
        }

        public Result<IEnumerable<UserResponse>> GetAll()
        {
            var users = _repository.GetAll();

            var response = users.Select(user => new UserResponse
            {
                Id = user.Id,
                Name = user.Name,
                CreatedAt = user.CreatedAt,
                UpdatedAt = user.UpdatedAt,
            });

            return Result<IEnumerable<UserResponse>>.Ok(response);
        }

        public Result<Unit> CreateName(string name)
        {
            // 이름 중복 검색
            if (_repository.FindByName(name) != null)
            {
                return Result<Unit>.Fail(Constants.DuplicateName);
            }

            var UtcNow = DateTime.UtcNow;

            var user = new UserEntity
            {
                Id = Guid.NewGuid(),
                Name = name,
                CreatedAt = UtcNow,
                UpdatedAt = UtcNow,
            };
            _repository.CreateName(user);
            return Result<Unit>.Ok(Unit.Value);
        }

        public Result<Unit> ChangeName(Guid id, string newName)
        {
            var user = _repository.FindById(id);
            if (user == null)
            {
                return Result<Unit>.Fail(Constants.UserNotFound);
            }

            // 이전 이름과 동일 이름
            if (user.Name == newName)
            {
                return Result<Unit>.Fail(Constants.DuplicateName);
            }

            // 이름 중복
            if (_repository.FindByName(newName) != null)
            {
                return Result<Unit>.Fail(Constants.DuplicateName);
            }

            user.Name = newName;
            user.UpdatedAt = DateTime.UtcNow;

            // DB 입력
            if (!_repository.ChangeName(user))
            {
                return Result<Unit>.Fail(Constants.DatabaseError);
            }

            return Result<Unit>.Ok(Unit.Value);
        }

        public Result<Unit> DeleteByIndex(int index)
        {
            var deleted = _repository.DeleteByIndex(index);
            return deleted
                ? Result<Unit>.Ok(Unit.Value)
                : Result<Unit>.Fail($"{Constants.InvalidIndex}: {index}");
        }

        public Result<Unit> DeleteByName(string name)
        {
            var count = _repository.DeleteByName(name);
            return count > 0
                ? Result<Unit>.Ok(Unit.Value)
                : Result<Unit>.Fail(Constants.NameNotFound);
        }
    }
}